src/NHibernate/Intercept/AbstractFieldInterceptor.cs

```diff
@@ -113,17 +114,22 @@ public object Intercept(object target, string fieldName, object value, bool sett
 					uninitializedFields.Remove(fieldName);
 				}
 
-				if (IsUninitializedAssociation(fieldName))
+				if (!unwrapProxyFieldNames.Contains(fieldName))
+				{
+					return value;
+				}
+
+				// When a proxy is set by the user which we know when the session is set, we should not unwrap it
+				if (session != null || !value.IsProxy() || NHibernateUtil.IsInitialized(value))
```

Comment:
Since on get, we assign from the proxy its unwrapped entity to the property in order to allow access after session closing, the `NHibernateUtil.IsInitialized(value)` condition should be removed in order to allow the unwrapping to take place even when the entity was proxy loaded prior to loading the parent entity having a `no-proxy` property on it.